Improve documentation of workflows and its steps (#1020)
workfloworchestrator/orchestrator-core: docs/reference-docs/workflows/workflow-steps.md

::: orchestrator.utils.state.inject_args
    options:
        heading_level: 3

> ## Reusable Workflow Steps in Orchestrator Core
> 
> When designing workflows in Orchestrator Core, reusability is key.

Reviewer: ```suggestion
## Reusable Workflow Steps

When designing workflows it is good practice to make steps reusable.
```

Don't think the "in Orchestrator Core" suffix is needed

For example, you may want to:

- Update a product's description using subscription properties.
- Push data to an external system via an upsert operation.

To take reusability further, Python's `@singledispatch` decorator can help abstract product-specific logic behind a common interface.
This makes your steps cleaner, easier to reuse across multiple workflows and more maintainable.

### Generic Workflow Steps

You can define a generic step when the logic should be shared across multiple workflows or products.
Here's an example of a reusable workflow step that updates an external system based on the product type:

```python
@step("Update external system")
def update_external_system(subscription: SubscriptionModel):
    match type(subscription):
        case ProductTypeOne:
            payload = create_product_type_one_payload(subscription.some_block)
        case ProductTypeTwo:
            payload = create_product_type_two_payload(subscription.some_other_block)
        case _:
            raise TypeError(f"Unsupported subscription type: {type(subscription)}")

    response = external_system_request(payload)
    return {"response": response}
```

> While this approach works, the switch logic (via match or if isinstance) can become unwieldy as more product types are introduced.

Reviewer: ```suggestion
While this approach works, the switch logic (via `match` or `if isinstance()`) can become unwieldy as more product types are introduced.
```
